File: Python3/801 - 1000/849 Maximize Distance to Closest Person.py
# 52ms, 86.87%; 13.5MB, 5.55%
class Solution:
    def maxDistToClosest(self, seats: list) -> int:
        beg, end = self.__find(seats), self.__find(seats, reverse=1)
        res = max(beg, len(seats) - end - 1)
        for i in range(beg + 1, end + 1):
            if seats[i] == 0: continue
            res = max(res, (i - beg) // 2)
            beg = i
        return res

    def __find(self, seats: list, reverse: int=0):
        if reverse == 0:
            beg, end, step = 0, len(seats), 1
        else:
            beg, end, step = len(seats) - 1, -1, -1
        for i in range(beg, end, step):
            if seats[i] == 1: return i
        return -1   # not needed

File: 801 - 1000/849 Maximize Distance to Closest Person.cc
// 8ms, 100.0%; 10.1MB, 98.80%
#include <bits/stdc++.h>
using namespace std;

static int x = []() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    return 0;
}();

class Solution {
public:
    int maxDistToClosest(const vector<int> &seats) {
        auto it1 = find(seats.begin(), seats.end(), 1);
        auto it2 = find(seats.rbegin(), seats.rend(), 1);

        int last = distance(seats.begin(), it1), end = distance(seats.rbegin(), it2);
        int res = max(last, end), limit = seats.size() - end;

        for (int i = last + 1; i < limit; ++i) {
            if (seats[i] == 0) continue;
            res = max(res, (i - last) / 2);
            last = i;
        }
        return res;
    }
};
